client/components/migration/migration-runner.js

```diff
@@ -0,0 +1,251 @@
+// to our state machine's state names. These are one-to-one mapping.
+// These state names are the key in the migrationStateTransitions object below.
+// Ref: Check state diagram in https://github.com/Automattic/woocommerce-services/pull/2757.
+const MIGRATION_ENUM_TO_STATE_NAME_MAP = {
+    2: 'stateInit',
+    3: 'stateErrorInit',
+    4: 'stateInstalling',
+    5: 'stateErrorInstalling',
+    6: 'stateActivating',
+    7: 'stateErrorActivating',
+    8: 'stateDB',
+    9: 'stateErrorDB',
+    10: 'stateDeactivating',
+    11: 'stateErrorDeactivating',
+    12: 'stateDone'
+}
+
+const plugins = 'woocommerce-shipping,woocommerce-tax'; //this needs to be a CSV string.
+const headers = {
+    "Content-Type": "application/json",
+    'X-WP-Nonce': getNonce()
+};
+
+const installPluginAPICall = () =>
+    fetch( getBaseURL() + 'wc-admin/plugins/install', {
+        method: 'POST',
+        headers,
+        body: JSON.stringify({
+            plugins
+        })
+    } );
+
+const activatePluginAPICall = () =>
+    fetch( getBaseURL() + 'wc-admin/plugins/activate', {
+        method: 'POST',
+        headers,
+        body: JSON.stringify({
+            plugins
+        })
+    } );
+
+const deactivateWCSTPluginAPICall = () =>
+    fetch( getBaseURL() + 'wp/v2/plugins/woocommerce-services/woocommerce-services', {
+        method: 'POST',
+        headers,
+        body: JSON.stringify({
+            status: 'inactive'
+        })
+    } );
+
+/**
+ * Note: this function can not be called under these conditions:
+ * 1. WCS&T, Woo Shipping, Woo Tax are all activated. WCS&T will not load the migration-flag controller, returning 404.
+ * 2. WCS&T is deactivated. The migration-flag controller is not loaded, this will return 404.
+ *
+ * @param {string} migrationState The numeric migration number from the PHP side. Check classes/class-wc-connect-wcst-to-wcshipping-migration-state-enum.php.
+ * @returns {Promise} Check fetch.
+ */
+const markMigrationStartedAPICall = (migrationState) => () =>
+    fetch( getBaseURL() + 'wc/v1/connect/migration-flag', {
+        method: 'POST',
+        headers,
+        body: JSON.stringify({
+            migration_state: migrationState
+        })
+    } );
+
+const stateErrorHandlingAPICall = (failedMigrationState) => () => {
+    // TODO: The error message doesn't show up anywhere. For now, update the flag in the option table so
+    // we know where it failed at.
+
+    return markMigrationStartedAPICall(failedMigrationState)();
+};
+
+/**
+ * A wrapper function to make API calls and check the API status response.
+ * Throw an exception up if the status > 400.
+ *
+ * @param {function} apiFn The API function that uses fetch.
+ * @returns {Object} The API JSON response from fetch.
+ */
+const fetchAPICall = (apiFn) => async () => {
+    const apiResponse = await apiFn();
+    const apiJSONResponse = await apiResponse.json();
```

Comment:
Good questions. In `migration-runner.js`, it doesn't make use of the response as long as it didn't return an error code. 

I changed the code 74e0998c a bit so that
- `apiResponse` is always a Response object because it's returned by `fetch`. https://developer.mozilla.org/en-US/docs/Web/API/fetch#return_value. We shouldn't need to worry about emptiness here. 
- `.json()` is only called when it's > 400 status. The error message will be in the body.
- Return the `fetch` `Response` object instead, this `Response` object can be `null` or a stream. https://developer.mozilla.org/en-US/docs/Web/API/Response/body#value. `migration-runner.js` doesn't consume the response so it won't break the logic.

Let me know what you think!

Reply:
sounds perfect, thanks!